File: src/providers/registry.ts
import consola from "consola"

import type { Provider, ProviderID } from "./types"

class ProviderRegistry {
  private providers = new Map<string, Provider>()
  private activeProvider: Provider | null = null
  private activeProviders = new Set<string>()

  async register(
    provider: Provider,
    options?: { saveConfig?: boolean },
  ): Promise<void> {
    this.providers.set(provider.instanceId, provider)
    if (options?.saveConfig === false) return
    try {
      const { saveConfig } = await import("~/lib/config-db")
      await saveConfig()
    } catch (err) {
      consola.warn("Failed to save config after provider registration:", err)
    }
  }

  getProvider(instanceId: string): Provider {
    const provider = this.providers.get(instanceId)
    if (!provider) {
      throw new Error(
        `Provider '${instanceId}' not found. Available: ${Array.from(this.providers.keys()).join(", ")}`,
      )
    }
    return provider
  }

  setActive(instanceId: string): Provider {
    const provider = this.getProvider(instanceId)
    this.activeProvider = provider
    this.activeProviders.add(instanceId)
    // Save config when activation state changes
    this.saveConfigAsync()
    return provider
  }

  getActive(): Provider {
    if (!this.activeProvider) {
      throw new Error("No active provider set")
    }
    return this.activeProvider
  }

  addActive(instanceId: string): Provider {
    const provider = this.getProvider(instanceId)
    this.activeProviders.add(instanceId)
    if (!this.activeProvider) {
      this.activeProvider = provider
    }
    // Save config when activation state changes
    this.saveConfigAsync()
    return provider
  }

  removeActive(instanceId: string): void {
    this.activeProviders.delete(instanceId)
    if (this.activeProvider?.instanceId === instanceId) {
      // Pick another active provider as the primary, or null
      const remaining = Array.from(this.activeProviders)
      this.activeProvider =
        remaining.length > 0 ? this.getProvider(remaining[0]) : null
    }
    // Save config when activation state changes
    this.saveConfigAsync()
  }

  getActiveProviders(): Array<Provider> {
    return Array.from(this.activeProviders)
      .filter((id) => this.providers.has(id))
      .map((id) => {
        const provider = this.providers.get(id)
        if (!provider) {
          throw new Error(`Active provider ${id} not found in registry`)
        }
        return provider
      })
  }

  isActiveProvider(instanceId: string): boolean {
    return this.activeProviders.has(instanceId)
  }

  listProviders(): Array<Provider> {
    return Array.from(this.providers.values())
  }

  isRegistered(instanceId: string): boolean {
    return this.providers.has(instanceId)
  }

  getProviderMap(): ReadonlyMap<string, Provider> {
    return this.providers
  }

  /** Re-register a provider under a new instanceId (e.g. after identity is known post-auth) */
  async rename(oldInstanceId: string, newInstanceId: string): Promise<boolean> {
    const provider = this.providers.get(oldInstanceId)
    if (!provider) return false

    // Check if target instanceId already exists
    if (this.providers.has(newInstanceId)) {
      consola.warn(
        `Cannot rename ${oldInstanceId} to ${newInstanceId}: target already exists`,
      )
      return false
    }

    this.providers.delete(oldInstanceId)
    this.providers.set(newInstanceId, provider)
    if (this.activeProviders.has(oldInstanceId)) {
      this.activeProviders.delete(oldInstanceId)
      this.activeProviders.add(newInstanceId)
    }
    if (this.activeProvider?.instanceId === newInstanceId) {
      // already updated since provider object mutated in place
    }
    // Save config after renaming
    try {
      const { saveConfig } = await import("~/lib/config-db")
      await saveConfig()
    } catch (err) {
      consola.warn("Failed to save config after provider rename:", err)
    }
    return true
  }

  /** Remove a provider instance and clean up its token file */
  async remove(instanceId: string): Promise<boolean> {
    const provider = this.providers.get(instanceId)
    if (!provider) return false

    // Remove from registry
    this.providers.delete(instanceId)

    // Remove from active providers
    if (this.activeProviders.has(instanceId)) {
      this.removeActive(instanceId)
    }

    // Clean up token file
    try {
      const { deleteTokenData } = await import("~/lib/token-db")
      await deleteTokenData(instanceId)
    } catch (err) {
      consola.warn(`Failed to clean up token for ${instanceId}:`, err)
    }

    // Save config after removal
    try {
      const { saveConfig } = await import("~/lib/config-db")
      await saveConfig()
    } catch (err) {
      consola.warn("Failed to save config after provider removal:", err)
    }

    return true
  }

  /** Return all registered instances of a given provider type */
  getInstancesOfType(id: ProviderID): Array<Provider> {
    return Array.from(this.providers.values()).filter((p) => p.id === id)
  }

  /** Generate a new unique instance ID for the given provider type */
  nextInstanceId(id: ProviderID): string {
    const existing = this.getInstancesOfType(id)
    // First instance keeps the plain ID, subsequent ones get unique suffixes
    if (existing.length === 0) return id

    // For providers that use username-based IDs (like GitHub Copilot),
    // we need to avoid conflicts with existing instances
    let counter = 2
    let candidateId: string
    do {
      candidateId = `${id}-${counter}`
      counter++
    } while (this.isRegistered(candidateId))

    return candidateId
  }

  /** Helper to save config asynchronously without blocking */
  private saveConfigAsync(): void {
    import("~/lib/config-db")
      .then(({ saveConfig }) => {
        saveConfig().catch((err: unknown) => {
          consola.warn(
            "Failed to auto-save config after provider state change:",
            err,
          )
        })
      })
      .catch((err: unknown) => {
        consola.warn("Failed to import config module:", err)
      })
  }
}

// Singleton instance
export const providerRegistry = new ProviderRegistry()

File: internal/registry/registry.go
// Package registry provides a centralized provider registry system
package registry

import (
	"encoding/json"
	"fmt"
	"omnillm/internal/database"
	"omnillm/internal/providers/types"
	"sync"

	"github.com/rs/zerolog/log"
)

type ProviderRegistry struct {
	mu              sync.RWMutex
	providers       map[string]types.Provider
	activeProvider  types.Provider
	activeProviders map[string]struct{}
	configStore     *database.ConfigStore
	instanceStore   *database.ProviderInstanceStore
}

var (
	globalRegistry *ProviderRegistry
	once           sync.Once
)

func GetProviderRegistry() *ProviderRegistry {
	once.Do(func() {
		globalRegistry = &ProviderRegistry{
			providers:       make(map[string]types.Provider),
			activeProviders: make(map[string]struct{}),
			configStore:     database.NewConfigStore(),
			instanceStore:   database.NewProviderInstanceStore(),
		}
	})
	return globalRegistry
}

func (pr *ProviderRegistry) Register(provider types.Provider, saveConfig bool) error {
	pr.mu.Lock()
	pr.providers[provider.GetInstanceID()] = provider
	shouldSave := saveConfig
	pr.mu.Unlock()

	if shouldSave {
		go pr.saveConfigAsync()
	}

	log.Debug().Str("provider", provider.GetInstanceID()).Msg("Provider registered")
	return nil
}

func (pr *ProviderRegistry) GetProvider(instanceID string) (types.Provider, error) {
	pr.mu.RLock()
	defer pr.mu.RUnlock()

	provider, exists := pr.providers[instanceID]
	if !exists {
		available := make([]string, 0, len(pr.providers))
		for id := range pr.providers {
			available = append(available, id)
		}
		return nil, fmt.Errorf("provider '%s' not found. Available: %v", instanceID, available)
	}
	return provider, nil
}

func (pr *ProviderRegistry) SetActive(instanceID string) (types.Provider, error) {
	pr.mu.Lock()
	provider, exists := pr.providers[instanceID]
	if !exists {
		pr.mu.Unlock()
		return nil, fmt.Errorf("provider '%s' not found", instanceID)
	}

	pr.activeProvider = provider
	pr.activeProviders[instanceID] = struct{}{}
	pr.mu.Unlock()

	go pr.saveConfigAsync()

	log.Debug().Str("provider", instanceID).Msg("Provider set as active")
	return provider, nil
}

func (pr *ProviderRegistry) GetActive() (types.Provider, error) {
	pr.mu.RLock()
	defer pr.mu.RUnlock()

	if pr.activeProvider == nil {
		return nil, fmt.Errorf("no active provider set")
	}
	return pr.activeProvider, nil
}

func (pr *ProviderRegistry) AddActive(instanceID string) (types.Provider, error) {
	pr.mu.Lock()
	provider, exists := pr.providers[instanceID]
	if !exists {
		pr.mu.Unlock()
		return nil, fmt.Errorf("provider '%s' not found", instanceID)
	}

	pr.activeProviders[instanceID] = struct{}{}
	if pr.activeProvider == nil {
		pr.activeProvider = provider
	}
	pr.mu.Unlock()

	go pr.saveConfigAsync()

	return provider, nil
}

func (pr *ProviderRegistry) RemoveActive(instanceID string) error {
	pr.mu.Lock()
	delete(pr.activeProviders, instanceID)
	if pr.activeProvider != nil && pr.activeProvider.GetInstanceID() == instanceID {
		// Pick another active provider as the primary, or nil
		pr.activeProvider = nil
		for id := range pr.activeProviders {
			if provider, exists := pr.providers[id]; exists {
				pr.activeProvider = provider
				break
			}
		}
	}
	pr.mu.Unlock()

	go pr.saveConfigAsync()

	return nil
}

func (pr *ProviderRegistry) GetActiveProviders() []types.Provider {
	pr.mu.RLock()
	defer pr.mu.RUnlock()

	var providers []types.Provider
	for id := range pr.activeProviders {
		if provider, exists := pr.providers[id]; exists {
			providers = append(providers, provider)
		}
	}
	return providers
}

func (pr *ProviderRegistry) IsActiveProvider(instanceID string) bool {
	pr.mu.RLock()
	defer pr.mu.RUnlock()

	_, exists := pr.activeProviders[instanceID]
	return exists
}

func (pr *ProviderRegistry) ListProviders() []types.Provider {
	pr.mu.RLock()
	defer pr.mu.RUnlock()

	providers := make([]types.Provider, 0, len(pr.providers))
	for _, provider := range pr.providers {
		providers = append(providers, provider)
	}
	return providers
}

func (pr *ProviderRegistry) IsRegistered(instanceID string) bool {
	pr.mu.RLock()
	defer pr.mu.RUnlock()

	_, exists := pr.providers[instanceID]
	return exists
}

func (pr *ProviderRegistry) GetProviderMap() map[string]types.Provider {
	pr.mu.RLock()
	defer pr.mu.RUnlock()

	// Return a copy to prevent external modifications
	result := make(map[string]types.Provider)
	for k, v := range pr.providers {
		result[k] = v
	}
	return result
}

func (pr *ProviderRegistry) Rename(oldInstanceID, newInstanceID string) error {
	pr.mu.Lock()
	provider, exists := pr.providers[oldInstanceID]
	if !exists {
		pr.mu.Unlock()
		return fmt.Errorf("provider '%s' not found", oldInstanceID)
	}

	if _, exists := pr.providers[newInstanceID]; exists {
		pr.mu.Unlock()
		log.Warn().Str("old", oldInstanceID).Str("new", newInstanceID).Msg("Cannot rename provider: target already exists")
		return fmt.Errorf("cannot rename %s to %s: target already exists", oldInstanceID, newInstanceID)
	}

	delete(pr.providers, oldInstanceID)
	pr.providers[newInstanceID] = provider

	if _, wasActive := pr.activeProviders[oldInstanceID]; wasActive {
		delete(pr.activeProviders, oldInstanceID)
		pr.activeProviders[newInstanceID] = struct{}{}
	}
	pr.mu.Unlock()

	go pr.saveConfigAsync()

	return nil
}

func (pr *ProviderRegistry) Remove(instanceID string) error {
	pr.mu.Lock()
	_, exists := pr.providers[instanceID]
	if !exists {
		pr.mu.Unlock()
		return fmt.Errorf("provider '%s' not found", instanceID)
	}

	delete(pr.providers, instanceID)
	delete(pr.activeProviders, instanceID)
	if pr.activeProvider != nil && pr.activeProvider.GetInstanceID() == instanceID {
		pr.activeProvider = nil
		for id := range pr.activeProviders {
			if p, exists := pr.providers[id]; exists {
				pr.activeProvider = p
				break
			}
		}
	}
	pr.mu.Unlock()

	// Clean up token file (outside lock)
	tokenStore := database.NewTokenStore()
	if err := tokenStore.Delete(instanceID); err != nil {
		log.Warn().Str("instance", instanceID).Err(err).Msg("Failed to clean up token")
	}

	go pr.saveConfigAsync()

	log.Debug().Str("provider", instanceID).Msg("Provider removed")
	return nil
}

func (pr *ProviderRegistry) GetInstancesOfType(providerType string) []types.Provider {
	pr.mu.RLock()
	defer pr.mu.RUnlock()

	var instances []types.Provider
	for _, provider := range pr.providers {
		if provider.GetID() == providerType {
			instances = append(instances, provider)
		}
	}
	return instances
}

func (pr *ProviderRegistry) NextInstanceID(providerType string) string {
	existing := pr.GetInstancesOfType(providerType)

	// First instance keeps the plain ID, subsequent ones get unique suffixes
	if len(existing) == 0 {
		return providerType
	}

	// For providers that use username-based IDs (like GitHub Copilot),
	// we need to avoid conflicts with existing instances
	counter := 2
	var candidateID string
	for {
		candidateID = fmt.Sprintf("%s-%d", providerType, counter)
		if !pr.IsRegistered(candidateID) {
			break
		}
		counter++
	}

	return candidateID
}

func (pr *ProviderRegistry) saveConfigAsync() error {
	// Snapshot data under lock to avoid holding it during I/O
	pr.mu.RLock()
	providers := make(map[string]types.Provider, len(pr.providers))
	for k, v := range pr.providers {
		providers[k] = v
	}
	activeProviders := make(map[string]struct{}, len(pr.activeProviders))
	for k, v := range pr.activeProviders {
		activeProviders[k] = v
	}
	pr.mu.RUnlock()

	// Save provider instances and activation state to database
	for id, provider := range providers {
		_, isActive := activeProviders[id]

		record := &database.ProviderInstanceRecord{
			InstanceID: id,
			ProviderID: provider.GetID(),
			Name:       provider.GetName(),
			Activated:  isActive,
		}

		// Preserve existing priority from DB if available
		if existing, err := pr.instanceStore.Get(id); err == nil && existing != nil {
			record.Priority = existing.Priority
		}

		if err := pr.instanceStore.Save(record); err != nil {
			log.Warn().Err(err).Str("instance", id).Msg("Failed to save provider instance")
		}
	}

	// Save active provider IDs to config store
	activeIDs := make([]string, 0, len(activeProviders))
	for id := range activeProviders {
		activeIDs = append(activeIDs, id)
	}

	activeJSON, _ := json.Marshal(activeIDs)
	if err := pr.configStore.Set("active_providers", string(activeJSON)); err != nil {
		log.Warn().Err(err).Msg("Failed to save active providers config")
	}

	log.Debug().Strs("active", activeIDs).Msg("Provider configuration saved")
	return nil
}
